services/data/src/hooks/useDataMutation.ts

```diff
@@ -0,0 +1,93 @@
+import { MutationState, Mutation, MutationRenderInput } from '../types/Mutation'
+import { useContext, useState, useEffect, useCallback, useRef } from 'react'
+import { DataContext } from '../components/DataContext'
+import { ContextType } from '../types/Context'
+import { QueryDefinition } from '../types/Query'
+import { joinPath } from '../utils/path'
+
+const mutationToQueryDef = (m: Mutation): QueryDefinition => ({
+    resource:
+        m.type === 'update' || m.type === 'delete'
+            ? joinPath(m.resource, m.id)
+            : m.resource,
+})
+
+const mutationToMethod = (m: Mutation): string => {
+    switch (m.type) {
+        case 'create':
+            return 'POST'
+        case 'update':
+            if (m.partial) {
+                return 'PATCH'
+            } else {
+                return 'POST'
+            }
+        case 'delete':
+            return 'DELETE'
+    }
+}
+
+const mutationToPayload = (m: Mutation): string | null => {
+    if (m.type === 'delete') {
+        return null
+    }
+    return JSON.stringify(m.data)
+}
+
+const fetchData = (
+    context: ContextType,
+    mutation: Mutation,
+    signal: AbortSignal
+) => {
+    const q = mutationToQueryDef(mutation)
+
+    return context.fetch(q, {
+        method: mutationToMethod(mutation),
+        body: mutationToPayload(mutation),
+        headers: {
+            'Content-Type': 'application/json',
+        },
+        signal: signal,
+    })
+}
+
+export const useDataMutation = (mutation: Mutation): MutationRenderInput => {
+    const context = useContext(DataContext)
+    const [state, setState] = useState<MutationState>({
+        loading: false,
+        called: false,
+    })
+
+    const controller = useRef<AbortController>()
+    const abort = () => {
+        controller.current && controller.current.abort()
+    }
+
+    const mutate = useCallback(() => {
+        if (controller.current) {
+            // Only ever mutate once!
```

Comment:
Hey @Mohammer5 !  I'm not sure I understand the question - if a component somewhere deeper in the `children` tree of the component using this hook re-renders, it shouldn't affect the top-level component at all (React is good at this).  If the **top-level** component re-renders, it will only update the `mutate` callback if the `context` or the `mutation` argument values change - see line 85.  This `if (controller.current) { return }; controller.current = new AbortController()` check isn't actually related to re-rendering at all, but rather to multiple calls to `mutate`.  This prevents a second call to `mutate` from duplicating the mutation.

Does that make sense?  Is there some way I could make it clearer in the code?

Reply:
One thing that I will have to double-check here is that this probably has a similar bug to #88 (though not as bad since it's not an infinite loop).  This again is problematic iff the `mutation` is created fresh on each render loop